lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php

```diff
@@ -43,47 +47,32 @@ class CompositeExpression implements Countable
      */
     public function __construct(string $type, array $parts = [])
     {
-        $this->type = $type;
-
-        $this->addMultiple($parts);
+        $this->type  = $type;
+        $this->parts = array_values(array_filter($parts, static function ($part) {
+            return ! ($part instanceof self && count($part) === 0);
+        }));
     }
 
     /**
-     * Adds multiple parts to composite expression.
+     * Returns a CompositeExpression with the given part added.
      *
-     * @param array<int, self|string> $parts
+     * This instance is immutable and unaffected by this method call.
      *
-     * @return $this
+     * @param self|string ...$parts
      */
-    public function addMultiple(array $parts = []) : self
+    public function with(...$parts) : self
     {
-        foreach ($parts as $part) {
-            $this->add($part);
-        }
+        $that = clone $this;
 
-        return $this;
-    }
-
-    /**
-     * Adds an expression to composite expression.
-     *
-     * @param self|string $part
-     *
-     * @return $this
-     */
-    public function add($part) : self
-    {
-        if (empty($part)) {
-            return $this;
-        }
+        foreach ($parts as $part) {
+            if ($part instanceof self && count($part) === 0) {
```

Comment:
Would it make sense to extract this into `private function isEmptySelf()` to generalize this logic and make its removal easier in the future?

Reply:
TBH I want to remove this code ASAP (i.e. in my next PR), so I don't think we should spend any more time attempting to make this code look prettier. It won't be easier to remove, either.

This "temporary" code is a small cost to pay to make one change at a time (I hesitated to resolve both #3844 and #3845 in the same PR, but decided not to, due to recent discussions).

Reply:
Given that we agreed to retarget this against 2.11, let's do that first. After then, it might be not possible to remove this code ASAP since it would require a breaking change.

Reply:
@BenMorel maybe I misinterpreted but I think @morozov is telling you that we can't have temporary code on 2.11.x because we want to be able to release 2.11.0 at any given time, and removing the temporary code after that would be a BC-break

Reply:
The extraction of this specific block to a private method won't be a BC break but this logic will have to stay in 2.x forever, so it still makes sense to generalize it.

Reply:
I'm not sure whether we're on the same page here, changes to internal code don't matter for BC, and I just meant I didn't want to spend too much time on stuff that is going to change in `master` again a few days later. But anyway, I've applied the suggestion in 2b01c576ebe2206683bceaad0c7d5f26f77850fd!